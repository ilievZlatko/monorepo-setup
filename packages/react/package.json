{
  "name": "@ds.zlatko/react",
  "version": "0.2.1",
  "main": "index.js",
  "license": "MIT",
  "files": [
    "lib"
  ],
  "devDependencies": {
    "@babel/core": "^7.17.9",
    "@babel/preset-env": "^7.16.11",
    "@babel/preset-react": "^7.16.7",
    "@babel/preset-typescript": "^7.16.7",
    "@storybook/addon-a11y": "^6.4.22",
    "@storybook/addon-actions": "^6.4.22",
    "@storybook/addon-essentials": "^6.4.22",
    "@storybook/addon-interactions": "^6.4.22",
    "@storybook/addon-knobs": "^6.4.0",
    "@storybook/addon-links": "^6.4.22",
    "@storybook/addon-storysource": "^6.4.22",
    "@storybook/preset-typescript": "^3.0.0",
    "@storybook/react": "^6.4.22",
    "@storybook/testing-library": "^0.0.9",
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.1.1",
    "@types/jest": "^27.4.1",
    "@types/react": "^18.0.3",
    "babel-loader": "^8.2.4",
    "chromatic": "^6.5.4",
    "jest": "^27.5.1",
    "react": "^18.0.0",
    "rollup": "^2.70.1",
    "rollup-plugin-typescript2": "^0.31.2",
    "typescript": "^4.6.3"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build --watch",
    "test": "jest --verbose",
    "test:watch": "yarn test --watch",
    "update:snapshots": "jest --updateSnapshot",
    "build-storybook": "build-storybook",
    "storybook": "start-storybook -p 6006",
    "chromatic": "chromatic --project-token=bcc8fa835e7b"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "0bcde58496d8214b2ed7314a522487660934f71c"
}
