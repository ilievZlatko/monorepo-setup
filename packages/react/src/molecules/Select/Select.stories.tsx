import React from "react";
import Select from "./Select";
import { withA11y } from "@storybook/addon-a11y";

// CSS
import "@dse.z/scss/lib/Select.css";

export default {
  title: "Molecules/Select",
  decorators: [withA11y],
};

const options = [
  {
    label: "Strict Black",
    value: "black",
  },
  {
    label: "Heavenly Green",
    value: "green",
  },
  {
    label: "Sweet Pink",
    value: "pink",
  },
];

export const Common = () => <Select options={options} />;

export const RenderOption = () => (
  <Select
    options={options}
    renderOption={({ getOptionRecommendedProps, option, isSelected }) => (
      <span {...getOptionRecommendedProps()}>
        {option.label} {isSelected ? "SELECTED" : ""}
      </span>
    )}
  />
);

export const CustomLabel = () => (
  <Select label="Select a color" options={options} />
);
