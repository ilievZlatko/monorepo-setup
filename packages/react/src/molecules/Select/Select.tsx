import React, {
  useState,
  useRef,
  createRef,
  useEffect,
  KeyboardEventHandler,
} from "react";
import Text from "../../atoms/Text";

interface SelectOption {
  label: string;
  value: string | null;
}

interface RenderOptionProps {
  isSelected: boolean;
  option: SelectOption;
  getOptionRecommendedProps: (overrideProps?: Object) => Object;
}

interface SelectProps {
  onOptionSelected?: (option: SelectOption, optionIndex: number) => void;
  options?: SelectOption[];
  label?: string;
  renderOption?: (props: RenderOptionProps) => React.ReactNode;
}

enum KEY_CODES {
  ENTER = 13,
  SPACE = 32,
  DOWN_ARROW = 40,
  ESC = 27,
  UP_ARROW = 38,
}

const getNextOptionIndex = (
  currentIndex: number | null,
  options: SelectOption[]
): number => {
  if (currentIndex === null) return 0;

  if (currentIndex === options.length - 1) return 0;

  return currentIndex + 1;
};

const getPrevOptionIndex = (
  currentIndex: number | null,
  options: SelectOption[]
): number => {
  if (currentIndex === null) return 0;

  if (currentIndex === 0) return options.length - 1;

  return currentIndex - 1;
};

const Select: React.FC<SelectProps> = ({
  onOptionSelected: handler,
  options = [],
  label = "Please select an option...",
  renderOption,
}) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [overlayTop, setOverlayTop] = useState<number>(0);
  const [selectedIndex, setSelectedIndex] = useState<number | null>(null);
  const [highlightedIndex, setHighlightedIndex] = useState<number | null>(null);
  const [optionRefs, setOptionRefs] = useState<
    React.RefObject<HTMLLIElement>[]
  >([]);
  const labelRef = useRef<HTMLButtonElement>(null);

  const onOptionSelected = (option: SelectOption, optionIndex: number) => {
    if (handler) handler(option, optionIndex);

    setSelectedIndex(optionIndex);
    setIsOpen(false);
  };

  const onLabelClick = () => {
    setIsOpen(!isOpen);
  };

  const highlightOption = (optionIndex: number | null) => {
    setHighlightedIndex(optionIndex);
  };

  const onButtonKeyDown: KeyboardEventHandler = (event) => {
    event.preventDefault();

    if (
      [KEY_CODES.DOWN_ARROW, KEY_CODES.ENTER, KEY_CODES.SPACE].includes(
        event.keyCode
      )
    ) {
      setIsOpen(true);
      highlightOption(0);
    }
  };

  const onOptionKeyDown: KeyboardEventHandler = (event) => {
    if (event.keyCode === KEY_CODES.ESC) {
      setIsOpen(false);
      return;
    }

    if (event.keyCode === KEY_CODES.DOWN_ARROW) {
      highlightOption(getNextOptionIndex(highlightedIndex, options));
    }

    if (event.keyCode === KEY_CODES.UP_ARROW) {
      highlightOption(getPrevOptionIndex(highlightedIndex, options));
    }

    if (event.keyCode === KEY_CODES.ENTER) {
      onOptionSelected(options[highlightedIndex!], highlightedIndex!);
    }
  };

  useEffect(() => {
    setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);
  }, [labelRef.current?.offsetHeight]);

  useEffect(() => {
    setOptionRefs(options.map(() => createRef<HTMLLIElement>()));
  }, [options.length]);

  useEffect(() => {
    if (highlightedIndex !== null && isOpen) {
      const ref = optionRefs[highlightedIndex];

      if (ref && ref.current) {
        ref.current.focus();
      }
    }
  }, [isOpen, highlightedIndex]);

  let selectedOption = null;

  if (selectedIndex !== null) {
    selectedOption = options[selectedIndex];
  }

  return (
    <div className="dse-select">
      <button
        data-testid="DseSelectButton"
        ref={labelRef}
        aria-haspopup="true"
        aria-expanded={isOpen ? true : undefined}
        aria-controls="dse-select-list"
        className="dse-select__label"
        onClick={onLabelClick}
        onKeyDown={onButtonKeyDown}
      >
        <Text>{selectedIndex === null ? label : selectedOption?.label}</Text>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className={`dse-select__caret ${
            isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"
          }`}
          width="1rem"
          height="1rem"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
          strokeWidth={2}
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            d="M19 9l-7 7-7-7"
          />
        </svg>
      </button>

      {
        <ul
          role="menu"
          id="dse-select-list"
          aria-hidden={isOpen ? undefined : false}
          style={{ top: overlayTop }}
          className={`dse-select__overlay ${
            isOpen ? "dse-select__overlay--open" : ""
          }`}
        >
          {options.map((option, optionIndex) => {
            const isSelected = selectedIndex === optionIndex;
            const isHighlighted = highlightedIndex === optionIndex;
            const ref = optionRefs[optionIndex];

            const renderOptionProps = {
              option,
              isSelected,
              getOptionRecommendedProps: (overrideProps = {}) => ({
                ref,
                role: "menuitemradio",
                "aria-label": option.label,
                "aria-checked": isSelected ? true : undefined,
                tabIndex: isHighlighted ? -1 : 0,
                onKeyDown: onOptionKeyDown,
                onMouseEnter: () => highlightOption(optionIndex),
                onMouseLeave: () => highlightOption(null),
                className: `dse-select__option ${
                  isSelected ? "dse-select__option--selected" : ""
                } ${isHighlighted ? "dse-select__option--highlighted" : ""}`,
                key: option.value,
                onClick: () => onOptionSelected(option, optionIndex),
                ...overrideProps,
              }),
            };

            if (renderOption) {
              return renderOption(renderOptionProps);
            }

            return (
              <li {...renderOptionProps.getOptionRecommendedProps()}>
                <Text>{option.label}</Text>

                {isSelected ? (
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    width="1rem"
                    height="1rem"
                    viewBox="0 0 20 20"
                    fill="currentColor"
                  >
                    <path
                      fillRule="evenodd"
                      d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
                      clipRule="evenodd"
                    />
                  </svg>
                ) : null}
              </li>
            );
          })}
        </ul>
      }
    </div>
  );
};

export default Select;
