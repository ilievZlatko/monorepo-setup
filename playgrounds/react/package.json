{
  "name": "@playgrounds/react",
  "version": "0.4.0",
  "main": "dist/index.js",
  "license": "MIT",
  "devDependencies": {
    "@types/react": "^18.0.3",
    "@types/react-dom": "^18.0.0",
    "parcel-bundler": "^1.12.5",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "typescript": "^4.6.3"
  },
  "scripts": {
    "dev": "parcel src/index.html -p 3000"
  },
  "private": true
}
