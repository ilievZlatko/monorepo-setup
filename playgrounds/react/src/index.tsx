import React from "react";
import ReactDOM from "react-dom";

import { Color } from "@ds.e/react/lib";
import "@ds.e/scss/lib/Utilities.css";

ReactDOM.render(
  <Color hexCode="#000" width="xl" height="xl" />,
  document.getElementById("root")
);

// .dse-width-xs  .dse-width-lg
