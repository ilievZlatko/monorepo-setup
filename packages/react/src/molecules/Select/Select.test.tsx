import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Select from "./Select";

const options = [
  {
    label: "Strict Black",
    value: "strict-black",
  },
  {
    label: "Heavenly Green",
    value: "heavenly-green",
  },
  {
    label: "Sweet Pink",
    value: "sweet-pink",
  },
];

test("renders all options passed to it", () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />);

  fireEvent.click(getByTestId("DseSelectButton"));
  expect(getAllByRole("menuitemradio")).toHaveLength(options.length);
});

test("renders options using custom renderOption method if passed as prop", () => {
  const { getAllByTestId, getByTestId } = render(
    <Select
      options={options}
      renderOption={({ option, getOptionRecommendedProps }) => (
        <li data-testid="CustomRenderOption" {...getOptionRecommendedProps()}>
          {option.label}
        </li>
      )}
    />
  );

  fireEvent.click(getByTestId("DseSelectButton"));

  expect(getAllByTestId("CustomRenderOption")).toHaveLength(options.length);
});

test("calls the onOptionSelected prop with the selected option and index", () => {
  const onOptionSelected = jest.fn();
  const { getAllByRole, getByTestId } = render(
    <Select options={options} onOptionSelected={onOptionSelected} />
  );

  fireEvent.click(getByTestId("DseSelectButton"));
  fireEvent.click(getAllByRole("menuitemradio")[0]);

  expect(onOptionSelected).toHaveBeenCalledWith(options[0], 0);
});

test("the button label changes on selected option", () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />);

  fireEvent.click(getByTestId("DseSelectButton"));
  fireEvent.click(getAllByRole("menuitemradio")[0]);

  expect(getByTestId("DseSelectButton")).toHaveTextContent(options[0].label);
});

test("snapshot of the selected option state", () => {
  const { getAllByRole, getByTestId, asFragment } = render(
    <Select options={options} />
  );

  fireEvent.click(getByTestId("DseSelectButton"));
  fireEvent.click(getAllByRole("menuitemradio")[0]);

  expect(asFragment()).toMatchSnapshot();
});

test("snapshot of the base state", () => {
  const { asFragment } = render(<Select options={options} />);

  expect(asFragment()).toMatchSnapshot();
});

test("snapshot of the options menu open state", () => {
  const { getByTestId, asFragment } = render(<Select options={options} />);

  fireEvent.click(getByTestId("DseSelectButton"));
  expect(asFragment()).toMatchSnapshot();
});

test("can customize select label", () => {
  const { getByText } = render(
    <Select options={options} label="THIS IS A CUSTOM LABEL" />
  );

  expect(getByText(/THIS IS A CUSTOM LABEL/)).toBeInTheDocument();
});
