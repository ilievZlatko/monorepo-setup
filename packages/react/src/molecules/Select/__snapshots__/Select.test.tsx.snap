// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Please select an option...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <ul
      aria-hidden="false"
      class="dse-select__overlay "
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Strict Black"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Strict Black
        </p>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Please select an option...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <ul
      class="dse-select__overlay dse-select__overlay--open"
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Strict Black"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Strict Black
        </p>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Strict Black
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <ul
      aria-hidden="false"
      class="dse-select__overlay "
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-checked="true"
        aria-label="Strict Black"
        class="dse-select__option dse-select__option--selected "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Strict Black
        </p>
        <svg
          fill="currentColor"
          height="1rem"
          viewBox="0 0 20 20"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option  "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
